torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -54,6 +56,95 @@ def reset_table(self) -> None:
 kernel_side_table = KernelSideTable()
 
 
+###############################################################################
+# Mutation Tracker
+
+
+@dataclasses.dataclass
+class MutationInfo:
+    mutated: bool = False
+    used_in_unknown: bool = False
+
+
+# Super basic mutation tracking pass that tracks which inputs are used in stores
+# It bails if any of the inputs are used in non tl.load/tl.store positions.
+# This pass will miss simple things like
+# a = in_ptr
+# tl.load(a, ...)
+# since it does not do any contextual analysis. This means that we might incorrectly
+# find extra mutations but this is safe as it would only be incorrect to miss
+# mutations.
+class MutationTracker(ast.NodeVisitor):
+    ALLOWED_READ_FNS = {
+        "load",
+        "max_constancy",
+        "max_contiguous",
+        "multiple_of",
+        "static_print",
+        "static_assert",
+        "device_print",
+        "device_assert",
+    }
+
+    def __init__(self, infos) -> None:
+        super().__init__()
+        self.infos = infos
+        self.read_depth = 0
+        self.in_store = False
+
+    def visit_Name(self, node):
+        if node.id not in self.infos:
+            return
+        if self.read_depth:
+            pass
+        elif self.in_store:
+            self.infos[node.id].mutated = True
+        else:
+            self.infos[node.id].used_in_unknown = True
+
+    def visit_Call(self, node):
+        # TODO(oulgen): Here we assume that there exists a line called
+        # from triton import language as tl. This needs to be checked
+        # as if someones imports xyz as tl then we will incorrectly
+        # assume a mutation but this would be ok as it is only unsafe to
+        # miss a mutation.
+        if (
+            isinstance(node.func, ast.Attribute)
+            and isinstance(node.func.value, ast.Name)
+            and node.func.value.id == "tl"
+        ):
+            if node.func.attr == "store":
+                # Do not allow for store to appear inside a read
+                # tl.load(a if tl.store(b) else z) is not useful
+                # and allowing this would complicate the analysis
+                assert self.read_depth == 0
+                assert self.in_store is False
+                self.in_store = True
+                self.generic_visit(node)
+                self.in_store = False
+                return
+            if node.func.attr in self.ALLOWED_READ_FNS:
+                self.read_depth += 1
+                self.generic_visit(node)
+                self.read_depth -= 1
+                return
+        self.generic_visit(node)
```

Comment:
Missed this one before: do we need to do `self.generic_visit(node)` again after the one in one of the `if` statements above? If not, should we make the whole thing `if ... else if ... else`?

Reply:
Scratch this: there are `return`s.